pallets/system/src/header_builder.rs

```diff
@@ -263,3 +264,207 @@ impl Decode for AvailExtrinsic {
+		}];
+
+		let data_root = data_root_avail(avail_ext);
+		let empty_data_root = data_root_avail(empty_data);
+		let single_data_root = data_root_avail(single_data);
+		//test for data root for appdata extrinsics
+		assert_eq!(data_root, [
+			221, 243, 104, 100, 122, 144, 42, 111, 106, 185, 245, 59, 50, 36, 91, 226, 142, 220,
+			153, 233, 47, 67, 240, 0, 75, 188, 44, 179, 89, 129, 75, 42
+		]);
+		//test for data root for empty appdata extrinsics
+		assert_eq!(empty_data_root, [0u8; 32]);
+		//test for data root for single extrinsic without appdata
+		assert_eq!(single_data_root, [0u8; 32]);
+	}
+
+	#[test]
+	fn test_merkle_proof() {
+		let mut leaves: Vec<[u8; 32]> = Vec::new();
+		let mut interested_leaf_position: Option<usize> = None;
+
+		let avail_data: Vec<Vec<u8>> = vec![
+			[
+				48, 51, 51, 49, 102, 97, 55, 51, 101, 101, 101, 99, 99, 98, 101, 52, 101, 50, 50,
+				53,
+			]
+			.to_vec(),
+			[
+				54, 48, 100, 101, 100, 49, 102, 53, 97, 98, 54, 55, 50, 97, 55, 49, 50, 55, 98, 97,
+			]
+			.to_vec(),
+			[
+				50, 98, 49, 49, 102, 49, 100, 100, 51, 57, 53, 53, 54, 102, 98, 50, 97, 98, 52, 50,
+			]
+			.to_vec(),
+		];
+
+		for (pos, xt) in avail_data.iter().enumerate() {
+			leaves.push(Sha256::hash(&xt));
+
+			if pos == 1usize {
+				interested_leaf_position = Some(leaves.len() - 1);
+			}
+		}
+
+		if leaves.len() > 0 {
+			if let Some(position) = interested_leaf_position {
+				let data_tree = MerkleTree::<Sha256>::from_leaves(&leaves);
+				let proof = data_tree.proof(&[position]);
+				let root_proof = proof.proof_hashes().to_vec();
+				assert_eq!(root_proof, vec![
+					[
+						117, 75, 148, 18, 224, 237, 121, 7, 189, 244, 183, 202, 93, 42, 34, 245,
+						225, 41, 160, 61, 235, 31, 78, 28, 31, 228, 45, 50, 47, 222, 233, 14
+					],
+					[
+						141, 110, 48, 228, 148, 209, 125, 118, 117, 169, 76, 60, 97, 68, 103, 255,
+						140, 206, 53, 32, 28, 16, 86, 117, 26, 110, 154, 16, 5, 21, 218, 249
+					]
+				]);
+			}
+		}
+	}
+
+	#[test]
+	fn test_single_merkle_proof() {
+		let mut leaves: Vec<[u8; 32]> = Vec::new();
+		let mut interested_leaf_position: Option<usize> = None;
+		let empty_vec: Vec<[u8; 32]> = vec![];
+
+		let avail_data: Vec<Vec<u8>> = vec![[
+			52, 53, 52, 102, 102, 56, 48, 99, 48, 56, 56, 97, 97, 55, 102, 97, 98, 57, 101, 49,
+		]
+		.to_vec()];
+
+		for (pos, xt) in avail_data.iter().enumerate() {
+			leaves.push(Sha256::hash(&xt));
+
+			if pos == 0usize {
+				interested_leaf_position = Some(leaves.len() - 1);
+			}
+		}
+		if leaves.len() > 0 {
+			if let Some(position) = interested_leaf_position {
+				let data_tree = MerkleTree::<Sha256>::from_leaves(&leaves);
+				let proof = data_tree.proof(&[position]);
+				let root_proof = proof.proof_hashes().to_vec();
+				// here the proof is shown empty because the root itself is the proof as there is only one appdata extrinsic
+				assert_eq!(root_proof, empty_vec);
+			}
+		}
+	}
+
+	#[test]
+	fn test_no_data_merkle_proof() {
```

Comment:
I'm not sure this test depends on any external code, it looks like it could be removed. Maybe we are missing some higher level abstraction, since part of a code is repeated in tests (`if leaves.len() > 0...`)